typed-protocols/src/Network/TypedProtocol/Proofs.hs

```diff
@@ -0,0 +1,184 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE TypeInType #-}
+
+
+-- | 
+--
+module Network.TypedProtocol.Proofs where
+
+import Network.TypedProtocol.Core
+import Network.TypedProtocol.Pipelined
+import Data.Void (Void, absurd)
+
+
+data AgencyProofs ps = AgencyProofs {
+
+       proofByContradiction_ClientAndServerHaveAgency
+         :: forall (st :: ps).
+            ClientHasAgency st
+         -> ServerHasAgency st
+         -> Void,
+
+       proofByContradiction_NobodyAndClientHaveAgency
+         :: forall (st :: ps).
+            NobodyHasAgency st
+         -> ClientHasAgency st
+         -> Void,
+
+       proofByContradiction_NobodyAndServerHaveAgency
+         :: forall (st :: ps).
+            NobodyHasAgency st
+         -> ServerHasAgency st
+         -> Void
+     }
+
+
+-- | The 'connect' function takes two peers that agree on a protocol and runs
+-- them in lock step, until (and if) they complete.
+--
+-- The 'connect' function serves a few purposes.
+--
+-- * The fact we can define this function at at all proves some minimal
+-- sanity property of the typed protocol framework.
+--
+-- * It demonstrates that all protocols defined in the framework can be run
+-- with synchronous communication rather than requiring buffered communication.
+--
+-- * It is useful for testing peer implementations against each other in a
+-- minimalistic setting. The typed framework guarantees
+--
+connect :: forall ps (st :: ps) m a b.
+           Monad m
+        => AgencyProofs ps
+        -> Peer ps AsClient st m a
+        -> Peer ps AsServer st m b
+        -> m (a, b)
```

Comment:
Should also give the terminal states.

Reply:
What is the benefit of that?

Reply:
If there are no terminal states, then you know that this `connect` can never finish.

Reply:
Beyond that, if you have more than one terminal state, you may wish to know which one the protocol reached.